homeassistant/components/guardian/switch.py

```diff
@@ -12,10 +21,88 @@
 
+SERVICE_UPGRADE_FIRMWARE_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_URL, default=DEFAULT_FIRMWARE_UPGRADE_URL): cv.url,
+        vol.Optional(CONF_PORT, default=DEFAULT_FIRMWARE_UPGRADE_PORT): cv.port,
+        vol.Optional(
+            CONF_FILENAME, default=DEFAULT_FIRMWARE_UPGRADE_FILENAME
+        ): cv.string,
+    }
+)
+
 
 async def async_setup_entry(hass, entry, async_add_entities):
     """Set up Guardian switches based on a config entry."""
+    _verify_domain_control = verify_domain_control(hass, DOMAIN)
+
     guardian = hass.data[DOMAIN][DATA_CLIENT][entry.entry_id]
+
+    @_verify_domain_control
+    async def disable_ap(call):
+        """Disable the device's onboard access point."""
+        try:
+            async with guardian.client:
+                await guardian.client.wifi.disable_ap()
+        except GuardianError as err:
+            LOGGER.error("Error during service call: %s", err)
+            return
+
+    @_verify_domain_control
+    async def enable_ap(call):
+        """Enable the device's onboard access point."""
+        try:
+            async with guardian.client:
+                await guardian.client.wifi.enable_ap()
+        except GuardianError as err:
+            LOGGER.error("Error during service call: %s", err)
+            return
+
+    @_verify_domain_control
+    async def reboot(call):
+        """Reboot the device."""
+        try:
+            async with guardian.client:
+                await guardian.client.system.reboot()
+        except GuardianError as err:
+            LOGGER.error("Error during service call: %s", err)
+            return
+
+    @_verify_domain_control
+    async def reset_valve_diagnostics(call):
+        """Fully reset system motor diagnostics."""
+        try:
+            async with guardian.client:
+                await guardian.client.valve.reset()
+        except GuardianError as err:
+            LOGGER.error("Error during service call: %s", err)
+            return
+
+    @_verify_domain_control
+    async def upgrade_firmware(call):
+        """Upgrade the device firmware."""
+        try:
+            async with guardian.client:
+                await guardian.client.system.upgrade_firmware(
+                    url=call.data[CONF_URL],
+                    port=call.data[CONF_PORT],
+                    filename=call.data[CONF_FILENAME],
+                )
+        except GuardianError as err:
+            LOGGER.error("Error during service call: %s", err)
+            return
+
+    platform = entity_platform.current_platform.get()
+
+    for service, method, schema in [
+        ("disable_ap", disable_ap, None),
+        ("enable_ap", enable_ap, None),
+        ("reboot", reboot, None),
+        ("reset_valve_diagnostics", reset_valve_diagnostics, None),
+        ("upgrade_firmware", upgrade_firmware, SERVICE_UPGRADE_FIRMWARE_SCHEMA),
+    ]:
+        platform.async_register_entity_service(service, schema, method)
```

Comment:
This doesn't work. `async_register_entity_service` has a different method signature for service handlers. 

Pass it a string and it will call that method on the switch entity.

Reply:
https://github.com/home-assistant/core/blob/dev/homeassistant/components/sonos/media_player.py#L256

Reply:
And if you do pass a function, the signature is `handle_service(entity, call)`.

Reply:
I didn't pay attention to the argument types. 😆